### PR TITLE
More API rework pending flushing logic fix

### PR DESCRIPTION
**Pull request by lu-zero**
Here some more refactors, some tests and a fix for a corner case condition.

Still in the way of fixing #1081 once for all.

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -489,35 +491,42 @@ impl Config {
       first_pass_data: FirstPassData {
         frames: Vec::new(),
       },
+      pool,
+      },
+      config: self.enc.clone(),
```
Comment (shssoichiro) [inline src/api.rs]: I had trouble reading this because of the indentation, could you please indent the inner struct contents?

File: src/api.rs [hunk 2]
```diff
@@ -489,35 +491,42 @@ impl Config {
...
+  pub(crate) config: EncoderConfig,
   rc_state: RCState,
   maybe_prev_log_base_q: Option<i64>,
   pub first_pass_data: FirstPassData,
   pool: rayon::ThreadPool,
 }
 
+pub struct Context<T: Pixel> {
+  inner: ContextInner<T>,
+  config: EncoderConfig,
```
Comment (shssoichiro) [inline src/api.rs]: What is the use case for having EncoderConfig in both Context and ContextInner? Is it avoidable somehow?

File: src/api.rs (hunk 1 above)
Comment (lu-zero) [reply src/api.rs]: Sure

File: src/api.rs (hunk 2 above)
Comment (lu-zero) [reply src/api.rs]: It is avoidable, but in the next set. Keeping it around reduces the amount of code changes.
